Knoema.Client/Client.cs

```diff
@@ -18,343 +19,540 @@
+			}
+			var builder = new UriBuilder(Uri.UriSchemeHttp, host)
+			{
+				Path = path,
+				Query = parameters != null ?
+					string.Join("&", parameters.Select(pair => string.Format("{0}={1}", pair.Key, HttpUtility.UrlEncode(pair.Value)))) :
+					string.Empty
+			};
+			return builder.Uri;
+		}
+
+		public Task<T> GetTaskResult<T>(int taskKey) where T : TaskResult
+		{
+			return ApiGet<T>("/api/1.0/meta/taskresult", string.Format("taskKey={0}", taskKey));
+		}
+
+		public async Task<T> WaitTaskResult<T>(int taskKey, int spinDelayInSeconds, int maxWaitCount) where T : TaskResult
+		{
+			T taskResult = null;
+			for (var i = 0; ;)
+			{
+				taskResult = await GetTaskResult<T>(taskKey);
+				if (!(taskResult.Status == Meta.TaskStatus.Executing || taskResult.Status == Meta.TaskStatus.Pending))
+					break;
+				i++;
+				if (i >= maxWaitCount)
+					throw new Exception("Maximum wait count reached");
+				Thread.Sleep(spinDelayInSeconds * 1000);
+			}
+
+			if (taskResult.Status == Meta.TaskStatus.Cancelled)
+				throw new TaskCanceledException("Task was cancelled");
+			if (taskResult.Status == Meta.TaskStatus.Failed)
+				throw new Exception(taskResult.Message);
+
+			if (taskResult.Status == Meta.TaskStatus.Completed)
+				return taskResult;
+
+			throw new ArgumentOutOfRangeException("Unexpected task status");
+		}
+
+		public Task<UnloadResponse> Unload(PivotRequest request)
+		{
+			return ApiPost<UnloadResponse>("/api/1.0/data/unload", request);
+		}
+
+		public async Task<DatasetUnloadTaskResultData> UnloadData(PivotRequest request, int spinDelayInSeconds = 10, int maxWaitCount = 360)
+		{
+			var unloadTask = await Unload(request);
+			var taskResult = await WaitTaskResult<DatasetUnloadTaskResult>(unloadTask.TaskKey, spinDelayInSeconds, maxWaitCount);
+			return taskResult.Data;
+		}
+
+		public Task<Task>[] UnloadGetFiles(string[] files, Stream[] resultStreams)
+		{
+			var cts = new CancellationTokenSource();
+			var clientHandler = new HttpClientHandler
+			{
+				AutomaticDecompression = DecompressionMethods.GZip | DecompressionMethods.Deflate,
+			};
+			var client = new HttpClient(clientHandler) { Timeout = TimeSpan.FromMilliseconds(HttpTimeout) };
+			var streamTasks = new Task<Task>[files.Length];
+
+			try
+			{
+				var getTasks = new Task<HttpResponseMessage>[files.Length];
+
+				for (var i = 0; i < files.Length; i++)
+				{
+					getTasks[i] = client.GetAsync(files[i], HttpCompletionOption.ResponseHeadersRead, cts.Token);
+				}
+
+				var contentTasks = new Task<HttpResponseMessage>[files.Length];
+				for (var i = 0; i < files.Length; i++)
+				{
+					contentTasks[i] = getTasks[i].ContinueWith(t =>
+					{
+						HttpResponseMessage response = null;
+						try
+						{
+							response = t.GetAwaiter().GetResult();
+							return response;
+						}
+						catch (Exception)
+						{
+							cts.Cancel();
+							if (response != null)
+								response.Dispose();
+
+							throw;
```

Comment:
Does the exception go to the top level (out of "ContinueWith" body)? If not, then you have to check the token for cancellation before next loop.

Reply:
it should, why not?